libcaf_core/caf/detail/ring_buffer.hpp

```diff
@@ -0,0 +1,95 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/config.hpp"
+
+#include <algorithm>
+#include <memory>
+
+namespace caf::detail {
+
+/// A simple ring buffer implementation.
+template <class T>
+class ring_buffer {
+public:
+  explicit ring_buffer(size_t max_size) : max_size_(max_size) {
+    if (max_size > 0)
+      buf_ = std::make_unique<T[]>(max_size);
+  }
+
+  ring_buffer(ring_buffer&& other) noexcept = default;
+
+  ring_buffer& operator=(ring_buffer&& other) noexcept = default;
+
+  explicit ring_buffer(const ring_buffer& other) {
```

Comment:
Copy constructors are never explicit:

```suggestion
  ring_buffer(const ring_buffer& other) {
```